sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+
+#### Device images collection
+
+DPC++ Runtime class named ProgramManager stores device images using following
+data structure:
+```
+/// Keeps all available device executable images added via \ref addImages.
+/// Organizes the images as a map from a kernel set id to the vector of images
+/// containing kernels from that set.
+/// Access must be guarded by the \ref Sync::getGlobalLock()
+std::unordered_map<SymbolSetId,
+                   std::unique_ptr<std::vector<RTDeviceBinaryImageUPtr>>>
+    m_DeviceImages;
+
+using StrToKSIdMap = std::unordered_map<string_class, SymbolSetId>;
+/// Maps names of kernels from a specific OS module (.exe .dll) to their set
+/// id (the sets are disjoint).
+std::unordered_map<OSModuleHandle, SymbolSetId> m_SymbolSets;
+```
+Assume each device image represents some combination of symbols and different
+device images may contain only exactly the same or not overlapping combination
+of symbols. If it is not so, there can be two cases:
+  - Symbols are the same. In this case it doesn't matter which device image is
+  taken to use duplicated symbol
+  - Symbols are not the same. In this case ODR violation takes place, such
+  situation leads to undefined behaviour. For more details refer to
+  [ODR violations](#ODR-violations) section.
+
+Each combination of symbols is assigned with an Id number - symbol set Id.
+A combination of symbols can exist in different formats (i.e. SPIR-V/AOT
+compiled binary and etc).
+`m_DeviceImages` maps an Id number to an array with device images which represent
+the same combination of symbols in different formats.
+`m_SymbolSets` contains mapping from symbol name to symbol set Id for each OS
+module (.exe/.so/.dll).
+`std::unordered_map` allows to search and access its elements with constant-time
+complexity.
+
+Before compilation of device image, to execute a kernel RT checks if the image
+contains any import information in its properties and if it does, then RT
+performs device images collection in order to resolve dependencies.
+
+Ids of all needed symbol sets are found by iterating through
+`m_SymbolSets` map, i.e. iterating through all available OS modules without
+predefined order and searching for first unresolved symbol in list of imports
+of target device image. Once device image that contains first symbol is
+met, remaining exported symbols are checked in found image. If
+they match some imported symbols then these matched symbols will be marked as
+resolved. The procedure repeats until all imported symbols are marked as
+resolved.
```

Comment:
What happens if no resolution could be found for an arbitrary symbol?

Reply:
I think in this case exception should be emitted. Or we can just pass unlinked module to the JIT compiler and it will an emit error.

Reply:
I believe, behaviour for such a case should be documented.

Reply:
Done.